policy-controller/src/profiling.rs

```diff
@@ -0,0 +1,144 @@
+        let lines: Vec<String> = self
+            .0
+            .data
+            .iter()
+            .map(|(key, value)| {
+                let mut line = key.thread_name_or_id();
+                line.push(';');
+
+                for frame in key.frames.iter().rev() {
+                    for symbol in frame.iter().rev() {
+                        write!(&mut line, "{};", symbol).unwrap();
+                    }
+                }
+
+                line.pop().unwrap_or_default();
+                write!(&mut line, " {}", value).unwrap();
+
+                line
+            })
+            .collect();
+
+        for line in lines {
+            writeln!(writer, "{}", line)?;
+        }
+
+        Ok(())
+    }
+
+}
+
+pub fn init() {
+    let guard = Arc::new(
+        ProfilerGuardBuilder::default()
+            .frequency(1000)
+            .blocklist(&["libc", "libgcc", "pthread", "vdso"])
+            .build()
+            .unwrap(),
+    );
+
+    let opt_query = warp::query::<QueryParams>()
+        .map(Some)
+        .or_else(|_| async { Ok::<(Option<QueryParams>,), std::convert::Infallible>((None,)) });
+
+    let pprof_report_endpoint =
+        warp::path("pprof")
+            .and(opt_query)
+            .map(move |params: Option<QueryParams>| {
+                let report = guard.report().build().unwrap(); 
+
+                match params {
+                    Some(obj) => {
+                        if obj.format == "flamegraph" {
+                            let mut file = Vec::new();
+                            report.flamegraph(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "image/svg+xml")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "proto" {
+                            let mut file = Vec::new();
+                            let profile = report.pprof().unwrap();
+                            profile.write_to_vec(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "application/octet-stream")
+                                .header("content-disposition", "attachment; filename=profile.pb")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "folded" {
+                            let my_report = MyReport(report);
+                            let mut file = Vec::new();
+                            my_report.folded(&mut file).unwrap();
+                            Response::builder()
+                                .header("content-type", "text/plain")
+                                .header("content-disposition", "attachment; filename=folded")
+                                .body(file)
+                                .unwrap()
+                        } else if obj.format == "tracing" {
```

Comment:
I think what you did in this branch belongs to the initialization phase, like when you called `ProfilerGuardBuilder::default()` above.